src/openlrr/game/world/Roof.cpp

```diff
@@ -22,30 +24,304 @@ LegoRR::Roof_Globs & LegoRR::roofGlobs = *(LegoRR::Roof_Globs*)0x00553180;
+										vertices[i].normal.x = worldPos.x + vertices[i].normal.x;
+										vertices[i].normal.y = worldPos.y + vertices[i].normal.y;
+										vertices[i].normal.z = worldPos.z + vertices[i].normal.z;
+									}
+								}
+								// Calc inverse then apply to normals
+								float InvTemp = 1.0f / sqrtf((vertices[i].normal.x * vertices[i].normal.x) + (vertices[i].normal.z * vertices[i].normal.z) + (vertices[i].normal.y * vertices[i].normal.y));
+								vertices[i].normal.z = InvTemp * vertices[i].normal.z;
+								vertices[i].normal.x = InvTemp * vertices[i].normal.x;
+								vertices[i].normal.y = InvTemp * vertices[i].normal.y;
+							}
+							Gods98::Container_Mesh_SetVertices(roofGlobs.contMesh, locRoofBlock->groupID, 0, 4, vertices);
+							/* &= ~ROOF_NEEDUPDATE; // (~0x2) updated, so clear flag */
+							locRoofBlock->flags &= ~ROOF_FLAG_NEEDUPDATE;
+							width = roofGlobs.dimensions.width;
+						}
+					}
+				}
+			}
+		}
+	}
+}
 
 // <LegoRR.exe @00463770>
-//void __cdecl LegoRR::Roof_Hide(bool32 hide);
+void __cdecl LegoRR::Roof_Hide(bool32 hide)
+{
+	// Check if roof is already hidden, if not, hide it
+	if (roofGlobs.hidden != hide) {
+		roofGlobs.hidden = hide;
+		if (hide != 0) {
+			Gods98::Container_Hide(roofGlobs.contMesh, TRUE);
+			return;
+		}
+		Container_Hide(roofGlobs.contMesh, FALSE);
+	}
+}
 
 // <LegoRR.exe @004637b0>
-//void __cdecl LegoRR::Roof_AddVisibleBlock(uint32 bx, uint32 by);
+void __cdecl LegoRR::Roof_AddVisibleBlock(uint32 bx, uint32 by)
+{
+	int idx;
+	
+	idx = (by * roofGlobs.dimensions.width) + bx;
+	Gods98::Container_Mesh_HideGroup(roofGlobs.contMesh, roofGlobs.grid[idx].groupID, FALSE);
+	roofGlobs.visibleTable[roofGlobs.visibleCount] = roofGlobs.grid + idx;
+	roofGlobs.visibleCount++;
+}
 
 // <LegoRR.exe @00463800>
-//void __cdecl LegoRR::Roof_HideAllVisibleBlocks(void);
+void __cdecl LegoRR::Roof_HideAllVisibleBlocks(void)
+{
+	// If any blocks visable clear them out
+	if (roofGlobs.visibleCount != 0) {
+		for(uint32 i = 0; i < roofGlobs.visibleCount; i++) {
+			Gods98::Container_Mesh_HideGroup(roofGlobs.contMesh, roofGlobs.visibleTable[i]->groupID, TRUE);
+		}
+	}
+	roofGlobs.visibleCount = 0;
+}
 
 // <LegoRR.exe @00463850>
-//bool32 __cdecl LegoRR::Roof_GetBlockPlaneNormal(uint32 bx, uint32 by, OUT Vector3F* normal);
+bool32 __cdecl LegoRR::Roof_GetBlockPlaneNormal(uint32 bx, uint32 by, OUT Vector3F* normal)
+{
+	float fVar1;
+	float fVar2;
+	float fVar3;
+	float fVar4;
+	Vector3F normal1;
+	Vector3F normal2;
+	Vertex vertices[4];
+
+	// Check if block is within bounds, calc normals, then assign to caller object
+	if (bx < (uint32)roofGlobs.dimensions.width && by < (uint32)roofGlobs.dimensions.height && (roofGlobs.grid[by * roofGlobs.dimensions.width + bx].flags & ROOF_FLAG_HIDDEN) == 0) {
+		Gods98::Container_Mesh_GetVertices(roofGlobs.contMesh,roofGlobs.grid[by * roofGlobs.dimensions.width + bx].groupID, 0, 4, vertices);
```

Comment:
Because a `RoofBlock` is being looked up twice here, it's easier to just look it up once and store the pointer to the block before the if statement. This will make the if statement less verbose. Also the flag operation can be switched to `!(roofBlock->flags & ROOF_FLAG_HIDDEN)`.